### ISSUE TITLE
Delayed first set of copies in traced region

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#1234]
I ran a 4-node 24-rank case on Summit. My app does explicit RK iterations. Each iteration is comprised of 3 algorithmically identical stages. Each stage has one round of peer-to-peer communication.

I am comparing two cases
1. the traced region between `begin_trace` and `end_trace` is 1 RK iteration ([profile](http://sapling.stanford.edu/~bandokihiro/Profile29/legion_prof))
2. the traced region between `begin_trace` and `end_trace` is 2 RK iterations ([profile](http://sapling.stanford.edu/~bandokihiro/Profile32/legion_prof))

As can be observed in the profiles, there is a gap on almost every rank only for the *first stage of the traced region*. Case (1) has a gap every 3 stages, case (2) every 6 stages. One initial hypothesis was that there is a slow rank for each of these gaps that is the source of the problem. However, looking only at rank 0, the channel starts to fire up, both for incoming but most importantly for outgoing transfers, 3-5 ms after all tasks that prepare the data on the producer side are completed. This is not the case for the other stages of a traced region.

I have just started looking at this with @streichler and we suspected something tracing related. Then case (2) is an experiment that was suggested to me by Alex. I think this confirms that the gap is tracing related but I am not sure if this is something expected and I can't identify in the profiles  what is causing this delay exactly.

**Comment by bandokihiro** [comment]
Some pictures extracted from the profiles
![Screenshot from 2022-04-14 17-03-03](https://user-images.githubusercontent.com/39278362/163496720-d9e4a319-e087-4ba4-b1fe-37383f66aae1.png)
![Screenshot from 2022-04-14 17-03-15](https://user-images.githubusercontent.com/39278362/163496725-fcff673b-91f8-4072-852a-15c63ffde9df.png)

**Comment by lightsighter** [comment]
It's not tracing related. Legion has issued trace replays way far into the future in this case. You can see all the "Trace Replay" tasks are gone from the utility processors starting around the 20th iteration or so. That means that Legion has built up the entire Realm event graph for the remainder of the execution of the program after only 20 iterations. There's nothing left for Legion to do at that point, it's all in Realm's hands. 

There are very strong indications that it is a data movement issue. For every rising edge of a "gap" in the utilization of the GPU on one processor, I can find a perfectly correlated falling edge in the utilization of the data movement channels on several other different nodes at exactly the same time. Those kinds of correlations are not coincidences, especially not when there is a repeated pattern of them. This just looks like jitter to me. Realm is dynamically executing event graphs completely independently on each node. There is no lockstep execution. Some nodes get ahead and then have to wait at communication points for slower nodes to catch up. It will appear non-deterministic and different nodes will manifest it from one run to the next. You can poke at it more if you want to by trying to use a memory allocator with more deterministic performance or by trying to make Realm's scheduling algorithms more deterministic, but tracing is definitely not the cause of the gaps.

**Comment by bandokihiro** [comment]
> For every rising edge of a "gap" in the utilization of the GPU on one processor, I can find a perfectly correlated falling edge in the utilization of the data movement channels on several other different nodes at exactly the same time.

If I understand correctly, this is indeed expected. A rank is wating for data to arrive and  we could actually find an incoming box on the channel of that same rank that ends exactly just before the next kernel kicks in. I think this doesn't explain why the channel starts ~4 ms after the data is ready to be sent. This shouldn't depend on slower neighboring ranks.

> This just looks like jitter to me. 
> It will appear non-deterministic 

I am not sure I understand. I feel like it is pretty deterministic occuring every x stages where x is the number of traced stages. I can even control how often these gaps occur. I agree that, based on what you described, the Legion's part of tracing seems unrelated. I am wondering if Realm does something special at the start of each replay which would delay things.

**Comment by alexaiken** [comment]
Mike,

Hmmm, it seems it is clearly tracing-related.  If you change the tracing, the behavior changes in a predictable way.  Maybe tracing isn't the direct cause, but then something connected to tracing is.

Alex
________________________________

**Comment by lightsighter** [comment]
> I think this doesn't explain why the channel starts ~4 ms after the data is ready to be sent. This shouldn't depend on slower neighboring ranks.

Sure it can. Anti-dependences are a real thing. Just because the data has been produced on one node doesn't mean that the destination instance finished being consumed. 

> I am not sure I understand. I feel like it is pretty deterministic occuring every x stages where x is the number of traced stages.
>Hmmm, it seems it is clearly tracing-related.  If you change the tracing, the behavior changes in a predictable way.  Maybe tracing isn't the direct cause, but then something connected to tracing is.

We haven't looked at any profiles yet without tracing yet and that is a bare minimum necessary condition to make me believe that this behavior is tracing related (it still won't be sufficient, but it is a first step). It could very easily be that the application/mapper behavior is changing as well and we haven't gone so far as to establish that yet.

If you want to actually convince me that it's all my fault then you need to dump the actual captured templates and point me at the unnecessary dependence. These traces look simple enough that you should be able to check them by hand.

**Comment by bandokihiro** [comment]
> dump the actual captured templates

which flag are you referring to? is it `-lg:dump_physical_traces`?

**Comment by bandokihiro** [comment]
Here is a [profile](http://sapling.stanford.edu/~bandokihiro/Profile33/legion_prof_0_1_2_6_7_8_9) with `-lg:no_tracing`. It is hard to conclude because there are gaps everywhere.

> Sure it can. Anti-dependences are a real thing.

I agree. Rank 9 is slowest [here](http://sapling.stanford.edu/~bandokihiro/Profile29/legion_prof). However, the anti-dependence is not systematic. At t=54,002 ms, rank 0 has finished packing its data to be sent. This is followed by the characteristic 3.5 ms delay of the first traced stage. However, all other ranks including rank 9 which is clearly the slowest have finished consuming the data from rank 0 of the previous stage.

**Comment by bandokihiro** [comment]
I put logs with `-lg:dump_physical_traces -level tracing=2` here `/home/bandokihiro/scratch/ForMike/logs`. I am not sure how to read those.

**Comment by lightsighter** [comment]
> However, the anti-dependence is not systematic

What do you mean by that?

> However, all other ranks including rank 9 which is clearly the slowest have finished consuming the data from rank 0 of the previous stage

That is definitely not true. Task 115785 from the previous iteration is still running on the GPU on rank 9 as late as 54.0277 seconds. I can't see which instances it's using, but I'd wager money there is at least one that needs to be copied into by the next iteration of the trace.

> I am not sure how to read those.

I'd recommend running with `-lg:parallel-replay 1` to get one slice per template, then they are easy to read. Look for the first (and probably only) "replayable" template. Then every line is of the form event[xxxxx] = instruction (events[yyyy]...). So you can see the dependences between operations. You're looking for a crossing event dependence from one iteration to the next that you believe shouldn't be there. If the traces are as simple as you claim there should only be a few instructions that you can analyze by hand.

**Comment by bandokihiro** [comment]
> Task 115785 from the previous iteration is still running on the GPU on rank 9 

Task 105816 on rank 0 and 115785 on rank 9 are part of the same index launch. The first stage of the traced region starts with task 105192 on rank 0 and task 115017 on rank 9 (dark blue). The last task from the previous stage on rank 9 that consumes data from rank 0 is task 114873. So rank 0 starts packing (tiny tasks after the dark blue task 105192) at 54,001 ms and the anti-dependence ceases to exist at 54,000 ms on rank 9.

> What do you mean by that?

I meant that the above example is a delay that cannot be explained by anti-dependence in my opinion.

@streichler also pointed out that all channels start at the same time there. There seems to be a global wait. On each rank, the channel starts when the utility procs show `Deferred Commit` with initiator `Trace Complete Operation` and `Trace Replay Operation`. You need to set "resolution=1" to be able to see them.

**Comment by elliottslaughter** [comment]
Has anyone checked for fence elision? If fence elision failed, that seems like it would potentially result in symptoms that look like these (i.e., different behavior at the start of a trace vs an iteration within the same trace).

**Comment by lightsighter** [comment]
> Has anyone checked for fence elision?

Fence elision doesn't fail. It can be turned off by the `-lg:no_fence_elision` or `-lg:no_trace_optimization` flags, but it will never "fail".

> On each rank, the channel starts when the utility procs show Deferred Commit with initiator Trace Complete Operation and Trace Replay Operation. You need to set "resolution=1" to be able to see them.

These are irrelevant to the execution of the Realm event graph. The graph has already been generated many seconds before we even get here. Once the `Replay Physical Trace` meta tasks disappear from the utility processors, then Legion has constructed the entire Realm event graph for the rest of the run. There is zero waiting by the Legion runtime for execution to finish in any of the profiles that you have sent.

> There seems to be a global wait.

I agree, but I'm saying it is caused by the application and not anything that Legion is doing. Legion's dependence analysis is precise. There are no spurious dependences and certainly no spurious dependences that span the entire machine. If there is a global wait it is because that is what the application has demanded from the semantics of the program that it executed.

What operations are you putting between the traces?

**Comment by bandokihiro** [comment]
> What operations are you putting between the traces?

Nothing relevant I think because I am disabling outputs and stuff for these performance runs. I very occasionally take time measurements but maybe 2-3 times during the whole run. This is especially the case for the run where I traced 2 RK iterations for which I commented most of the stuff between traces. Besides incrementing the iteration counter, I really don't do much. Nothing stands out from the profiles between traces as well and I expected that something would show up if I really was triggering a global wait.

**Comment by elliottslaughter** [comment]
> Fence elision doesn't fail.

I thought that fence elision can fail at least if there are any tasks or operations between the end of one trace and the start of the next?

**Comment by bandokihiro** [comment]
So I don't know what fence elision is about but with `-lg:no_fence_elision`, the gap is placed right at the trace boundaries.

**Comment by lightsighter** [comment]
> So I don't know what fence elision is about

Section 4F: https://legion.stanford.edu/pdfs/trace2018.pdf

> Nothing relevant 
> I thought that fence elision can fail at least if there are any tasks or operations between the end of one trace and the start of the next?

Trace elision doesn't fail. Putting operations in between your traces though does mean that the runtime must disable the fence elision optimization to be sound and that will put global barriers between your traces executions and the intermediate operations when running with control replication (with only a few strictly controlled exceptions). That's not failing; it is necessary for the correctness of execution.

**Comment by bandokihiro** [comment]
I left a reproducer with a 2-rank case that should produce logging which shows an unexpected dependency. The associated legion is at `/home/bandokihiro/scratch/ForMike/Issue1234/Repro/legion`

If you wanna make changes to legion, rebuild with the followings (on a gpu node)
```
cd /home/bandokihiro/scratch/ForMike/Issue1234/Repro/legion/build
make install
cd /home/bandokihiro/scratch/ForMike/Issue1234/Repro/DG-Legion/build
make solver
```

To run the case, do
```
cd /home/bandokihiro/scratch/ForMike/Issue1234/Repro/Run
sbatch runslurm.sh
```
This will produce logging with `-lg:dump_physical_traces -lg:parallel_replay 1 -level tracing=2` in a `logs` subdirectory.

**Comment by lightsighter** [comment]
This looks like a very old tracing bug in the fence elision. I need Wonchan to confirm my understanding of his code before I can push a fix.

If what I think is the bug turns out to not be the bug, then we are going to need a simpler example. Traces with several thousands of instructions cannot be analyzed by hand. This code is completely deterministic and you should be able to reproduce the issue with a trace of no more than 100 instructions. Right now I don't even know where to begin to look for the problem. Everything here is completely deterministic and repeatable and therefore should be easily reducible using standard delta debugging techniques.

**Comment by bandokihiro** [comment]
The above reproducer will generate in `/home/bandokihiro/scratch/ForMike/Issue1234/Repro/Run/logs` log files where there are only 70 events in the replayable section. The pattern is a follow

```
[0 - 7f27200c6000]    0.783789 {2}{tracing}: #### Replayable 0x7f21b817d260 Trace 1 for top_level_task (UID 6) ####

(...)

[0 - 7f27200c6000]    0.783915 {2}{tracing}:   events[77] = Runtime::create_ap_user_event()    (owner: (5,0))
[0 - 7f27200c6000]    0.783918 {2}{tracing}:   Runtime::trigger_event(events[77], events[0])    (owner: (5,0))

(...)

[0 - 7f27200c6000]    0.783926 {2}{tracing}:   events[83] = Runtime::merge_events(events[69],events[77],events[81],events[234],events[243])    (owner: (5,0))
[0 - 7f27200c6000]    0.783935 {2}{tracing}:   events[84] = copy(operations[(5,0)], Index expr: 146, {(400040004080000a,0,8,200000,0),(400040004080000a,0,8,200001,0),(400040004080000a,0,8,200002,0),(400040004080000a,0,8,200003,0)}, {(400000000080001d,0,8,200000,0),(400000000080001d,0,8,200001,0),(400000000080001d,0,8,200002,0),(400000000080001d,0,8,200003,0)}, events[83])
```

**Comment by lightsighter** [comment]
Pull the current control replication branch and see if it is gone.

**Comment by bandokihiro** [comment]
The profile looks happy now with gaps being really explained by unbalance between ranks. However, the pattern in the logs didn't disappear. I wanted to make sure this is expected before closing the issue.

**Comment by lightsighter** [comment]
That is expected. This was a very old tracing bug, that existed since the very first incarnation of tracing.

**Comment by bandokihiro** [comment]
All right. Thank you for the fix.
